File: engine/wal_manifest_test.go
// Copyright (c) 2026 Kiruba Sankar Swaminathan
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root of this source tree.

package engine

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

func testManifest(id uint32, file string) *walManifest {
	return &walManifest{
		Version:     walManifestVersion,
		SegmentSize: 4096,
		ActiveID:    id,
		Segments: []walManifestSegment{{
			ID:      id,
			File:    file,
			BaseLSN: 0,
		}},
	}
}

func TestSaveWalManifest_AtomicReplace(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, walManifestName)

	if err := saveWalManifest(path, testManifest(1, "seg-000001.wal")); err != nil {
		t.Fatal(err)
	}
	if _, err := os.Stat(walManifestTmpPath(path)); !os.IsNotExist(err) {
		t.Fatal("expected no leftover tmp after a successful save")
	}

	if err := saveWalManifest(path, testManifest(2, "seg-000002.wal")); err != nil {
		t.Fatal(err)
	}
	got, err := loadWalManifest(path)
	if err != nil {
		t.Fatal(err)
	}
	if got.ActiveID != 2 || got.Segments[0].File != "seg-000002.wal" {
		t.Fatalf("loaded %+v, want active 2 / seg-000002.wal", got)
	}
}

func TestSaveWalManifest_TornTmpKeepsOld(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, walManifestName)
	if err := saveWalManifest(path, testManifest(1, "seg-000001.wal")); err != nil {
		t.Fatal(err)
	}

	// Crash during the next write: dest is complete, tmp is garbage.
	if err := os.WriteFile(walManifestTmpPath(path), []byte("{"), fileMode); err != nil {
		t.Fatal(err)
	}
	got, err := loadWalManifest(path)
	if err != nil {
		t.Fatal(err)
	}
	if got.ActiveID != 1 {
		t.Fatalf("ActiveID=%d, want 1 (torn tmp must not replace dest)", got.ActiveID)
	}
}

func TestLoadWalManifest_PromotesCompleteTmp(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, walManifestName)
	tmp := walManifestTmpPath(path)

	data, err := json.MarshalIndent(testManifest(3, "seg-000003.wal"), "", "  ")
	if err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(tmp, append(data, '\n'), fileMode); err != nil {
		t.Fatal(err)
	}

	got, err := loadWalManifest(path)
	if err != nil {
		t.Fatal(err)
	}
	if got.ActiveID != 3 {
		t.Fatalf("ActiveID=%d, want 3 from recovered tmp", got.ActiveID)
	}
	if _, err := os.Stat(path); err != nil {
		t.Fatalf("expected dest after promoting tmp: %v", err)
	}
	if _, err := os.Stat(tmp); !os.IsNotExist(err) {
		t.Fatal("tmp should be gone after promote")
	}
}

func TestLoadWalManifest_TornTmpWithoutDest(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, walManifestName)
	if err := os.WriteFile(walManifestTmpPath(path), []byte("{"), fileMode); err != nil {
		t.Fatal(err)
	}
	_, err := loadWalManifest(path)
	if !os.IsNotExist(err) {
		t.Fatalf("torn tmp without dest: %v, want ErrNotExist", err)
	}
}

func TestOpenDataLog_RecoversManifestTmp(t *testing.T) {
	dir := t.TempDir()
	log, err := OpenDataLog(dir, nil, 4096)
	if err != nil {
		t.Fatal(err)
	}
	payload := encodeRecord(Record{Type: RecordSet, XID: 1, Key: []byte("k"), Value: []byte("v")})
	if _, err := log.AppendEncoded(payload, true); err != nil {
		t.Fatal(err)
	}
	if err := log.Close(); err != nil {
		t.Fatal(err)
	}

	walDir := filepath.Join(dir, walDirName)
	path := filepath.Join(walDir, walManifestName)
	tmp := walManifestTmpPath(path)
	if err := os.Rename(path, tmp); err != nil {
		t.Fatal(err)
	}

	log2, err := OpenDataLog(dir, nil, 4096)
	if err != nil {
		t.Fatal(err)
	}
	defer log2.Close()
	if _, err := os.Stat(path); err != nil {
		t.Fatalf("open should promote tmp to dest: %v", err)
	}
	if log2.SegmentCount() != 1 {
		t.Fatalf("segments=%d, want 1 after tmp recover", log2.SegmentCount())
	}
}

File: rust/crates/turnstone-engine/tests/wal_manifest_test.rs
// Copyright (c) 2026 Kiruba Sankar Swaminathan
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root of this source tree.

use std::fs;
use std::io;
use std::path::PathBuf;

use turnstone_engine::wal::{
    load_wal_manifest, save_wal_manifest, wal_manifest_tmp_path, WalManifest, WalManifestSegment,
    WAL_MANIFEST_VERSION,
};

fn test_manifest(id: u32, file: &str) -> WalManifest {
    WalManifest {
        version: WAL_MANIFEST_VERSION,
        segment_size: 4096,
        scan_floor: 0,
        active_id: id,
        segments: vec![WalManifestSegment {
            id,
            file: file.to_string(),
            base_lsn: 0,
            end_lsn: 0,
        }],
    }
}

#[test]
fn save_wal_manifest_atomic_replace() {
    let dir = tempfile::tempdir().unwrap();
    let path = dir.path().join("manifest.json");

    save_wal_manifest(&path, &test_manifest(1, "seg-000001.wal")).unwrap();
    assert!(fs::metadata(wal_manifest_tmp_path(&path)).is_err());

    save_wal_manifest(&path, &test_manifest(2, "seg-000002.wal")).unwrap();
    let got = load_wal_manifest(&path).unwrap();
    assert_eq!(got.active_id, 2);
    assert_eq!(got.segments[0].file, "seg-000002.wal");
}

#[test]
fn save_wal_manifest_torn_tmp_keeps_old() {
    let dir = tempfile::tempdir().unwrap();
    let path = dir.path().join("manifest.json");
    save_wal_manifest(&path, &test_manifest(1, "seg-000001.wal")).unwrap();
    fs::write(wal_manifest_tmp_path(&path), b"{").unwrap();
    let got = load_wal_manifest(&path).unwrap();
    assert_eq!(got.active_id, 1);
}

#[test]
fn load_wal_manifest_promotes_complete_tmp() {
    let dir = tempfile::tempdir().unwrap();
    let path = dir.path().join("manifest.json");
    let tmp = wal_manifest_tmp_path(&path);
    let data = serde_json::to_vec_pretty(&test_manifest(3, "seg-000003.wal")).unwrap();
    let mut data = data;
    data.push(b'\n');
    fs::write(&tmp, data).unwrap();

    let got = load_wal_manifest(&path).unwrap();
    assert_eq!(got.active_id, 3);
    assert!(path.exists());
    assert!(fs::metadata(&tmp).is_err());
}

#[test]
fn load_wal_manifest_torn_tmp_without_dest() {
    let dir = tempfile::tempdir().unwrap();
    let path = dir.path().join("manifest.json");
    fs::write(wal_manifest_tmp_path(&path), b"{").unwrap();
    let err = load_wal_manifest(&path).unwrap_err();
    assert_eq!(err.kind(), io::ErrorKind::NotFound);
}

#[test]
fn open_data_log_recovers_manifest_tmp() {
    let dir = tempfile::tempdir().unwrap();
    let log = turnstone_engine::DataLog::open(dir.path(), 4096).unwrap();
    let payload = turnstone_engine::encode_record(&turnstone_engine::Record {
        ty: turnstone_engine::RecordType::Set,
        xid: 1,
        key: b"k".to_vec(),
        value: b"v".to_vec(),
    });
    log.append_encoded(&payload, true).unwrap();
    log.close().unwrap();

    let wal_dir: PathBuf = dir.path().join("wal");
    let path = wal_dir.join("manifest.json");
    let tmp = wal_manifest_tmp_path(&path);
    fs::rename(&path, &tmp).unwrap();

    let log2 = turnstone_engine::DataLog::open(dir.path(), 4096).unwrap();
    assert!(path.exists());
    assert_eq!(log2.segment_count(), 1);
    log2.close().unwrap();
}
